lib/sdf-server/src/service/v2/func/create_func.rs

```diff
@@ -151,14 +249,39 @@ pub async fn create_func(
                 } else {
                     inputs.into_iter().map(|input| input.into()).collect()
                 };
-                FuncAuthoringClient::create_new_leaf_func(
+                let func = FuncAuthoringClient::create_new_leaf_func(
                     &ctx,
                     request.name,
                     LeafKind::CodeGeneration,
                     EventualParent::SchemaVariant(schema_variant_id),
                     &inputs,
                 )
-                .await?
+                .await?;
+                ctx.write_audit_log(
+                    AuditLogKind::CreateFunc {
+                        func_display_name: func.display_name.clone(),
+                        func_kind: func.kind.into(),
+                    },
+                    func.name.clone(),
+                )
+                .await?;
+                let schema_variant_name =
+                    SchemaVariant::get_by_id_or_error(&ctx, schema_variant_id)
+                        .await?
+                        .display_name()
+                        .to_string();
+                ctx.write_audit_log(
+                    AuditLogKind::AttachCodeGenFunc {
+                        func_id: func.id,
+                        func_display_name: func.display_name.clone(),
+                        schema_variant_id: Some(schema_variant_id),
+                        component_id: None,
+                        subject_name: schema_variant_name,
+                    },
+                    func.name.clone(),
+                )
```

Comment:
Right now we create audit logs outside of the `dal`, but I'm wondering if `CreateFunc` should just move to `Func::new`. Then again, some of the `new` methods are run by other `copy` or `clone` or `duplicate` methods... so the audit log doesn't always reveal the user intent...

Just rambling on your PR, but I'm thinking about making this less manual in some way. Feels like there's room for a crate in between the `dal` low level and the route authoring that could exist.

Reply:
yeah I had similar thoughts - it's also notable we usually end up with 3 things per route, ws event(s), posthog events, and now audit events and usually the posthog and audit events are quite similar. Not sure what to do with it exactly, but it crossed my mind as I was in here.

Reply:
Yep... not this PR but it's starting to bubble over.